From ebab6bdb9fed14628663f079566c76f7666a129d Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:29 -0400
Subject: [powerpc] powernv: Fix little endian issues in OPAL error log code

Message-id: <1410545655-205645-461-git-send-email-dzickus@redhat.com>
Patchwork-id: 94431
O-Subject: [RHEL7 PATCH 460/626] powerpc/powernv: Fix little endian issues in OPAL error log code
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 14ad0c58d5df6e5911a5413abdc2a9be6a8acb51
Author: Anton Blanchard <anton@samba.org>
Date:   Tue Apr 22 15:01:25 2014 +1000

    powerpc/powernv: Fix little endian issues in OPAL error log code

    Fix little endian issues with the OPAL error log code.

    Signed-off-by: Anton Blanchard <anton@samba.org>
    Reviewed-by: Stewart Smith <stewart@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/opal.h b/arch/powerpc/include/asm/opal.h
index cb7d52e..1a752ac 100644
--- a/arch/powerpc/include/asm/opal.h
+++ b/arch/powerpc/include/asm/opal.h
@@ -859,7 +859,7 @@ int64_t opal_lpc_read(uint32_t chip_id, enum OpalLPCAddressType addr_type,
         uint32_t addr, __be32 *data, uint32_t sz);
 
 int64_t opal_read_elog(uint64_t buffer, uint64_t size, uint64_t log_id);
-int64_t opal_get_elog_size(uint64_t *log_id, uint64_t *size, uint64_t *elog_type);
+int64_t opal_get_elog_size(__be64 *log_id, __be64 *size, __be64 *elog_type);
 int64_t opal_write_elog(uint64_t buffer, uint64_t size, uint64_t offset);
 int64_t opal_send_ack_elog(uint64_t log_id);
 void opal_resend_pending_logs(void);
diff --git a/arch/powerpc/platforms/powernv/opal-elog.c b/arch/powerpc/platforms/powernv/opal-elog.c
index fa195d4..b823afb 100644
--- a/arch/powerpc/platforms/powernv/opal-elog.c
+++ b/arch/powerpc/platforms/powernv/opal-elog.c
@@ -243,18 +243,25 @@ static struct elog_obj *create_elog_obj(uint64_t id, size_t size, uint64_t type)
 
 static void elog_work_fn(struct work_struct *work)
 {
+ __be64 size;
+ __be64 id;
+ __be64 type;
  uint64_t elog_size;
  uint64_t log_id;
  uint64_t elog_type;
  int rc;
  char name[2+16+1];
 
- rc = opal_get_elog_size(&log_id, &elog_size, &elog_type);
+ rc = opal_get_elog_size(&id, &size, &type);
  if (rc != OPAL_SUCCESS) {
   pr_err("ELOG: Opal log read failed\n");
   return;
  }
 
+ elog_size = be64_to_cpu(size);
+ log_id = be64_to_cpu(id);
+ elog_type = be64_to_cpu(type);
+
  BUG_ON(elog_size > OPAL_MAX_ERRLOG_SIZE);
 
  if (elog_size >= OPAL_MAX_ERRLOG_SIZE)
-- 
1.7.1